controllers/nstemplateset/namespaces.go

```diff
@@ -352,7 +386,7 @@ func getNamespaceName(request reconcile.Request) (string, error) {
 
 // isUpToDateAndProvisioned checks if the obj has the correct Template Reference Label.
 // If so, it processes the tier template to get the expected roles and rolebindings and then checks if they are actually present in the namespace.
-func (r *namespacesManager) isUpToDateAndProvisioned(ctx context.Context, ns *corev1.Namespace, tierTemplate *tierTemplate) (bool, error) {
+func (r *namespacesManager) isUpToDateAndProvisioned(ctx context.Context, ns *corev1.Namespace, tierTemplate *tierTemplate, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
```

Comment:
The current implementation for the namespaced feature-enabled resources is limited to RoleBindings and Roles only (as the only namespaced resources we are currently truly restoring if deleted). And unless I'm missing something the PR already covers the case of enabling features for existing NSTemplateSets.

What was missing is proper feature-disabling for existing NSTemplateSets. There was a bug in the `isUpToDateAndProvisioned()` function. I actually had a test for that specific use case but there was a bug in the test too so both bugs gave me a false negative result :)
This bug should be fixed here: https://github.com/codeready-toolchain/member-operator/pull/580/commits/3679b4ce0cdaedc8a5bb64369857ea5f0cae5eb0
I also extended the tests to cover both cases better.

Basically what the current code (in master) does: it always checks if the existing Roles &RBs match what is in the templates even if there is no changes in the reference (last applied templates). The current PR just adds an additional check if they also match the desired NSTemplate features. And unless I'm missing something there is no need for storing last applied features for that.

Alternative solution would be, as you said, storing the information about last applied features AND filtering out objects for disabled features in `tierTemplate.process()` (Rajiv's idea). That could potentially allows us to support all namespaced objects. Not just Roles and RBs.
I tried to avoid adding more annotations like "last-applied-features" and take advantage of our existing Role & RoleBinding handling. I hoped that a simple extension to the current Role & RB handling logic would be trivial but it turned out to be more complicated than expected. So I will give it a try and see if it's worth it.

Reply:
ok, so in case of namespace-scoped resources, you want to support feature-enabled resources for Roles/RoleBindings only for now, is that correct? And if the feature-enabled resources don't contain either Role or RoleBinding, then the controller will not apply the change for the already-existing NSTemplateSets.

I'm fine with that, let's just create a story to make sure that this is handled properly when we are doing the refactoring of the NSTemplateSet controller. cc @mfrancisc

Reply:
This also means that the changes of the feature-toggles won't be reflected if the resources are defined in the SpaceRole templates. We update the SpaceRole templates only if there is some change between:
1. `toolchain.dev.openshift.com/last-applied-space-roles` annotation
2. current content of `NSTemplateSet.Spec.SpaceRoles`

Reply:
That being said, since the support of the feature-enabled namespace-scoped resources is very limited in your PR and since we need to support only ClusterRoleBinding to be able to enable OLS in DevSandbox, don't you want to simplify the code and support cluster-scoped resources only for now?

We can take care of the broad/generic support of the feature-enabled resources when we will work on the changes of the NSTemplateSet controller. We will need to store the "last updated features" in status and re-apply everything if it's not matching with the current one in the annotation.

Reply:
I deleted all changes for namespaced resources. The PR supports the cluster scoped resource only now.